backbone.js

```diff
@@ -1140,6 +1140,7 @@
         removed.push(model);
         this._removeReference(model, options);
       }
+      if (models.length > 0 && !options.silent) delete options.index;
```

Comment:
If you expose something as part of an API, you shouldn't constrain how people look at what was exposed regardless of hypothetical use-cases.  I think most people would choose consistency over inconsistency if consistency comes almost for free.

In terms of cost:

N = num elements to delete
The performance hit to `_removeModels` of **1 more delete** in comparison to the existing **N splices** in `_removeModels` + **2N deletes** in `_removeReference` is insignificant.

All 3 of the solutions below fix the bug that the ***WRONG*** `options.index `is being sent to the 'add' event.  I don't have a strong opinion about which solution to implement, I went with the most conservative because it seemed like it would have the least downstream side-effects and the cost really is minimal.  The `delete` happens once per collection `set` IFF `options.remove:true` AND `options.silent:false` AND the collection has elements to remove. 

**Solutions**
* `delete options.index` - **consistent**  _cases 1,3_ This is the truest representation of the documentation/changelog "`options.index` will no longer be set ", and has the least side-effects to options.  You are more likely to leave options in the state that it was in when it was passed into `set`.  If we are trying to minimize side-effects, then we should add a `delete options.index` after it is used by the 'add' callbacks as well.
* `options.index = void *;`  and `setOptions` has a default `index:void 0` option -  **consistent** _cases 1,2_
* `options.index = void *;` - **inconsistent** _cases 1,2,3_